File: engine/src/flutter/lib/web_ui/lib/src/engine/canvaskit/image_filter.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:typed_data';

import 'package:ui/src/engine.dart';
import 'package:ui/ui.dart' as ui;

abstract class CkImageFilter implements BackendImageFilter {
  /// Returns the Skia native filter.
  ///
  /// Returns null if the filter is invalid (e.g. non-invertible matrix) or if
  /// it's a no-op (e.g. an erode/dilate with a radius of 0).
  SkImageFilter? get nativeFilter => _skImageFilter;
  SkImageFilter? _skImageFilter;

  @override
  void dispose() {
    _skImageFilter?.delete();
    _skImageFilter = null;
  }

  @override
  ui.Rect filterBounds(ui.Rect inputBounds) {
    final SkImageFilter? skFilter = nativeFilter;
    if (skFilter == null) {
      return inputBounds;
    }
    final Int32List? outputBounds = skFilter.getOutputBounds(toSkRect(inputBounds));
    // The CanvasKit C++ API may return null if the output bounds cannot be
    // computed. In such cases we fall back to the input bounds.
    if (outputBounds == null) {
      return inputBounds;
    }
    return rectFromSkIRect(outputBounds);
  }
}

class CkColorFilterImageFilter extends CkImageFilter {
  CkColorFilterImageFilter(BackendColorFilter colorFilter) {
    _skImageFilter = canvasKit.ImageFilter.MakeColorFilter(
      (colorFilter as CkColorFilter).skiaObject,
      null,
    );
  }
}

class CkBlurImageFilter extends CkImageFilter {
  CkBlurImageFilter({
    required double sigmaX,
    required double sigmaY,
    required ui.TileMode tileMode,
  }) {
    _skImageFilter = canvasKit.ImageFilter.MakeBlur(sigmaX, sigmaY, toSkTileMode(tileMode), null);
  }
}

class CkMatrixImageFilter extends CkImageFilter {
  CkMatrixImageFilter({required Float64List matrix, required ui.FilterQuality filterQuality}) {
    _skImageFilter = canvasKit.ImageFilter.MakeMatrixTransform(
      toSkMatrixFromFloat64(matrix),
      toSkFilterOptions(filterQuality),
      null,
    );
  }
}

class CkDilateImageFilter extends CkImageFilter {
  CkDilateImageFilter({required double radiusX, required double radiusY}) {
    _skImageFilter = canvasKit.ImageFilter.MakeDilate(radiusX, radiusY, null);
  }
}

class CkErodeImageFilter extends CkImageFilter {
  CkErodeImageFilter({required double radiusX, required double radiusY}) {
    _skImageFilter = canvasKit.ImageFilter.MakeErode(radiusX, radiusY, null);
  }
}

class CkComposeImageFilter extends CkImageFilter {
  CkComposeImageFilter({required BackendImageFilter outer, required BackendImageFilter inner}) {
    final SkImageFilter? skOuter = (outer as CkImageFilter).nativeFilter;
    final SkImageFilter? skInner = (inner as CkImageFilter).nativeFilter;
    _skImageFilter = canvasKit.ImageFilter.MakeCompose(skOuter, skInner);
  }
}

File: impeller/toolkit/interop/image_filter.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "impeller/toolkit/interop/image_filter.h"

namespace impeller::interop {

ImageFilter::ImageFilter(std::shared_ptr<const flutter::DlImageFilter> filter)
    : filter_(std::move(filter)) {}

ImageFilter::~ImageFilter() = default;

ScopedObject<ImageFilter> ImageFilter::MakeBlur(Scalar x_sigma,
                                                Scalar y_sigma,
                                                flutter::DlTileMode tile_mode) {
  auto filter = flutter::DlBlurImageFilter::Make(x_sigma, y_sigma, tile_mode);
  if (!filter) {
    return nullptr;
  }
  return Create<ImageFilter>(std::move(filter));
}

ScopedObject<ImageFilter> ImageFilter::MakeDilate(Scalar x_radius,
                                                  Scalar y_radius) {
  auto filter = flutter::DlDilateImageFilter::Make(x_radius, y_radius);
  if (!filter) {
    return nullptr;
  }
  return Create<ImageFilter>(std::move(filter));
}

ScopedObject<ImageFilter> ImageFilter::MakeErode(Scalar x_radius,
                                                 Scalar y_radius) {
  auto filter = flutter::DlErodeImageFilter::Make(x_radius, y_radius);
  if (!filter) {
    return nullptr;
  }
  return Create<ImageFilter>(std::move(filter));
}

ScopedObject<ImageFilter> ImageFilter::MakeMatrix(
    const Matrix& matrix,
    flutter::DlImageSampling sampling) {
  auto filter =
      flutter::DlMatrixImageFilter::Make(ToSkMatrix(matrix), sampling);
  if (!filter) {
    return nullptr;
  }
  return Create<ImageFilter>(std::move(filter));
}

ScopedObject<ImageFilter> ImageFilter::MakeCompose(const ImageFilter& outer,
                                                   const ImageFilter& inner) {
  auto filter = flutter::DlComposeImageFilter::Make(outer.GetImageFilter(),
                                                    inner.GetImageFilter());
  if (!filter) {
    return nullptr;
  }
  return Create<ImageFilter>(std::move(filter));
}

const std::shared_ptr<const flutter::DlImageFilter>&
ImageFilter::GetImageFilter() const {
  return filter_;
}

}  // namespace impeller::interop
